core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -927,27 +928,43 @@ private void visitAllEntities(Solution_ solution, Consumer<Object> visitor,
         }
     }
 
-    public List<Object> getEntityListByEntityClass(Solution_ solution, Class<?> entityClass) {
-        List<Object> entityList = new ArrayList<>();
+    public void visitEntitiesByEntityClass(Solution_ solution, Class<?> entityClass, Consumer<Object> visitor) {
         for (MemberAccessor entityMemberAccessor : entityMemberAccessorMap.values()) {
-            if (entityMemberAccessor.getType().isAssignableFrom(entityClass)) {
+            if (entityClass.isAssignableFrom(entityMemberAccessor.getType())) {
                 Object entity = extractMemberObject(entityMemberAccessor, solution);
-                if (entity != null && entityClass.isInstance(entity)) {
-                    entityList.add(entity);
+                if (entity != null) {
+                    visitor.accept(entity);
                 }
             }
         }
         for (MemberAccessor entityCollectionMemberAccessor : entityCollectionMemberAccessorMap.values()) {
-            // TODO if (entityCollectionPropertyAccessor.getPropertyType().getElementType().isAssignableFrom(entityClass)) {
-            Collection<Object> entityCollection = extractMemberCollectionOrArray(entityCollectionMemberAccessor, solution,
-                    false);
-            for (Object entity : entityCollection) {
-                if (entityClass.isInstance(entity)) {
-                    entityList.add(entity);
+            Optional<Class<?>> optionalTypeParameter = ConfigUtils.extractCollectionGenericTypeParameterLeniently(
+                    "solutionClass", entityCollectionMemberAccessor.getDeclaringClass(),
+                    entityCollectionMemberAccessor.getType(),
+                    entityCollectionMemberAccessor.getGenericType(),
+                    null,
+                    entityCollectionMemberAccessor.getName());
+            if (optionalTypeParameter.isPresent()) {
+                // In a typical case, typeParameter is specified, so we can skip the collection if typeParam
+                // is not assignable to entityClass.
+                Class<?> typeParameter = optionalTypeParameter.get();
+                if (entityClass.isAssignableFrom(typeParameter)) {
+                    Collection<Object> entityCollection =
+                            extractMemberCollectionOrArray(entityCollectionMemberAccessor, solution, false);
+                    entityCollection.forEach(visitor);
+                }
```

Comment:
This branch implements the TODO on line 941. If the collection has a type parameter and it is a subtype of the entity class, then visit all elements.